R/types.R

```diff
@@ -431,10 +431,89 @@ switch_class <- function(.x, ...) {
 coerce_class <- function(.x, .to, ...) {
   switch(class(.x), ..., abort_coercion(.x, .to))
 }
-abort_coercion <- function(x, to) {
-  abort(paste0(
-    "Cannot convert objects of type `", type_of(x), "` to `", to, "`"
-  ))
+abort_coercion <- function(x, to_type) {
+  x_type <- friendly_type(type_of(x))
+  if (!inherits(to_type, "AsIs")) {
+    to_type <- friendly_type(to_type)
+  }
+  abort(paste0("Cannot convert ", x_type, " to ", to_type))
+}
+
+friendly_type <- function(type) {
+  friendly <- friendly_type_of(type)
+  if (!is_null(friendly)) {
+    return(friendly)
+  }
+
+  friendly <- friendly_lang_type_of(type)
+  if (!is_null(friendly)) {
+    return(friendly)
+  }
+
+  friendly <- friendly_expr_type_of(type)
+  if (!is_null(friendly)) {
+    return(friendly)
+  }
+
+  type
+}
+
+friendly_type_of <- function(type) {
+  switch(type,
+    logical = "a logical vector",
+    integer = "an integer vector",
+    numeric = ,
+    double = "a double vector",
+    complex = "a complex vector",
+    character = "a character vector",
+    raw = "a raw vector",
+    string = "a string",
+    list = "a list",
+
+    NULL = "the NULL object",
+    environment = "an environment",
+    externalptr = "a pointer",
+    weakref = "a weak reference",
+    S4 = "an S4 object",
+
+    name = ,
+    symbol = "a symbol",
+    language = "a language call",
+    pairlist = "a pairlist node",
+    expression = "an expression vector",
+    quosure = "a quosure",
+
+    char = "an internal string",
+    promise = "an internal promise",
+    ... = "an internal dots object",
+    any = "an internal `any` object",
+    bytecode = "an internal bytecode object",
+
+    primitive = ,
+    builtin = ,
+    special = "a primitive function",
+    closure = "a function"
+  )
+}
+
+friendly_lang_type_of <- function(type) {
+  switch(type,
+    named = "a named language call",
```

Comment:
I think I'd prefer to make the connection to language when we can so that people can understand the prefix `lang_`. Parentheses looks like a good option.